### PR TITLE
Split into Raw and "safe"

### PR DESCRIPTION
**Pull request by miniBill**

### CODE REVIEW

File: src/Stream.elm [hunk 1]
```diff
@@ -1,653 +1,304 @@
...
+isEmpty (Stream _ f) =
+    Raw.isEmpty f
 
-        Immature thunk ->
-            isEmpty (thunk ())
 
-        _ ->
-            False
+equals : Stream a -> Stream a -> Outcome Bool
+equals (Stream ll l) (Stream rl r) =
```
Comment (z5h) [inline src/Stream.elm:215]: If one is finite and another is infinite, we know for certain they aren't equal.

File: src/Stream.elm [hunk 2]
```diff
@@ -1,653 +1,304 @@
...
+    case l of
+        Finite ->
+            Known False
 
+        Infinite ->
+            Known True
 
+        Indefinite ->
+            -- `length` is a bogus function to force the stream
+            Unsafe (\_ -> Raw.length s < 0)
```
Comment (z5h) [inline src/Stream.elm:266]: I think `Unsafe (\_ -> Raw.length s |> always False` is more "correct".

File: src/Stream.elm [hunk 3]
```diff
@@ -1,653 +1,304 @@
...
-            else
-                member a s_
+        Infinite ->
+            Known False
 
-        Immature thunk ->
-            member a (thunk ())
+        Indefinite ->
+            -- `length` is a bogus function to force the stream
+            Unsafe (\_ -> Raw.length s >= 0)
```
Comment (z5h) [inline src/Stream.elm:252]: I think `Unsafe (\_ -> Raw.length s |> always True` is more "correct".

**Review by z5h — APPROVED** [review]
One correction, and two preferences. Otherwise looks good.

File: src/Stream.elm (hunk 1 above)
Comment (miniBill) [reply src/Stream.elm:215]: True!
Also, if either is finite, calling `equals` is always safe!

**Pull request merged**